csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -56,24 +45,80 @@ class AllocationOrderInferencer : public IterVisitor {
   void handle(BinaryOp*) override;
   void handle(TernaryOp*) override;
   void handle(PadOp*) override;
+  void handle(ReductionOp*) override;
   // TODO: Add more propagation rules
-  // void handle(Reduction*) override;
   // void handle(LoadStoreOp*) override;
   // void handle(SqueezeOp*) override;
   // void handle(ExpandOp*) override;
 
  private:
-  // propagate allocation order from src to dst. Returns true when src has a
-  // recorded allocation order, false otherwise.
-  bool propagateAllocationOrder(TensorView* src, TensorView* dst) {
-    if (auto iter = alloc_order_map_.find(src);
+  // mapping allocation domain from producer to consumer without reduction
+  //
+  // e.g.
+  //   producer rfactor dom [r0', i0', i1', i2'] @ allocation order {0, 1, 3, 2}
+  //    |       alloc dom [r0', i0', i2', i1']
+  //    |
+  //    Operation
+  //    |
+  //    v
+  //   consumer rfactor dom [..., i0, ..., i1, ..., i2, ...]
+  //
+  // we construct allocation domain on producer, filtering out reduction, apply
+  // root domain map from producer to consumer.
+  //   [r0', i0', i2', i1'] -> [i0', i2', i1'] -> [i0, i2, i1]
+  // so the function would return [i0, i2, i1]
+  std::vector<IterDomain*> propagateAllocationDomain(
+      TensorView* producer,
+      TensorView* consumer) {
+    // constructing alloc_domain for producer from its root domain, while
+    // filtering out reduction because they won't appear in consumer's domain.
+    std::vector<IterDomain*> alloc_domain = TensorDomain::noReductions(
+        constructAllocationDomain(producer, alloc_order_map_.at(producer)));
+    // creating producer to consumer root domain map
+    std::unordered_map<IterDomain*, IterDomain*> p2c_map =
+        PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
+    // map alloc_domain to consumer
+    std::transform(
+        alloc_domain.cbegin(),
+        alloc_domain.cend(),
+        alloc_domain.begin(),
+        [&p2c_map](IterDomain* id) { return p2c_map.at(id); });
```

Comment:
Note that the root domain map may not map all domains.

Reply:
Actually, I don't seem to understand how this would work. `alloc_domain` is effectively a (potentially reordered) list of iter domains of the producer rfactor domain. Why could that be the allocation domain of the consumer? For example, how would reshape work?